core/src/main/java/org/bitcoinj/core/LegacyP2PKHAddress.java

```diff
@@ -0,0 +1,140 @@
+ */
+public class LegacyP2PKHAddress extends LegacyAddress {
+
+    public static final Script.ScriptType OUTPUT_SCRIPT_TYPE = Script.ScriptType.P2PKH;
+
+    private LegacyP2PKHAddress(NetworkParameters params, byte[] hash160) throws AddressFormatException {
+        super(params, hash160);
+    }
+
+    /**
+     * Construct a {@link LegacyP2PKHAddress} that represents the given pubkey hash. The resulting address will be a P2PKH type of
+     * address.
+     *
+     * @param params
+     *            network this address is valid for
+     * @param hash160
+     *            20-byte pubkey hash
+     * @return constructed address
+     */
+    public static LegacyP2PKHAddress fromPubKeyHash(NetworkParameters params, byte[] hash160) throws AddressFormatException {
+        return new LegacyP2PKHAddress(params, hash160);
+    }
+
+    /**
+     * Construct a {@link LegacyP2PKHAddress} that represents the public part of the given {@link ECKey}. Note that an address is
+     * derived from a hash of the public key and is not the public key itself.
+     *
+     * @param params
+     *            network this address is valid for
+     * @param key
+     *            only the public part is used
+     * @return constructed address
+     */
+    public static LegacyP2PKHAddress fromKey(NetworkParameters params, ECKey key) {
+        return fromPubKeyHash(params, key.getPubKeyHash());
+    }
+
+    /**
+     * Construct a {@link LegacyP2PKHAddress} from its base58 form.
+     *
+     * @param params
+     *            expected network this address is valid for, or null if if the network should be derived from the
+     *            base58
+     * @param base58
+     *            base58-encoded textual form of the address
+     * @throws AddressFormatException
+     *             if the given base58 doesn't parse or the checksum is invalid
+     * @throws AddressFormatException.WrongAddressType
+     *             if the given base58 corresponds to a P2SH address
+     * @throws AddressFormatException.WrongNetwork
+     *             if the given address is valid but for a different chain (eg testnet vs mainnet)
+     */
+    public static LegacyP2PKHAddress fromBase58(@Nullable NetworkParameters params, String base58)
+            throws AddressFormatException, AddressFormatException.WrongAddressType, AddressFormatException.WrongNetwork {
+        byte[] versionAndDataBytes = Base58.decodeChecked(base58);
+        int version = versionAndDataBytes[0] & 0xFF;
+        byte[] bytes = Arrays.copyOfRange(versionAndDataBytes, 1, versionAndDataBytes.length);
+        if (params == null) {
+            for (NetworkParameters p : Networks.get()) {
+                if (version == p.getAddressHeader())
+                    return fromPubKeyHash(p, bytes);
+                if (version == p.getP2SHHeader())
+                    throw new AddressFormatException.WrongAddressType("P2PKH");
```

Comment:
Shouldn't this say "P2SH"? Same 7 lines below.

Reply:
I've changed the error message to try to be clearer.